Introduce platform-agnostic `ALWAYS_INLINE` macro

hebasto opened the pull request:
Split from https://github.com/bitcoin/bitcoin/pull/24773 as requested in https://github.com/bitcoin/bitcoin/pull/24773#issuecomment-1534954977.

theuni commented on src/attributes.h:
```diff
@@ -16,4 +16,12 @@
+#elif defined(_MSC_VER)
+#  define ALWAYS_INLINE __forceinline
+#else
+#  define ALWAYS_INLINE inline
```
I think it'd be better to error in this case so that we can add another platform exception if necessary. IIRC for the places that we use it, inlining is critical and we might want to design functions differently for platforms where it's not possible.

Especially given the [gcc semantics](https://gcc.gnu.org/onlinedocs/gcc/Common-Function-Attributes.html):

> Failure to inline such a function is diagnosed as an error.

As-is, I think this is currently a theoretical regression as we might now compile on a platform that would be an error on master.

hebasto replied:
Mind suggesting an error wording?

theuni replied:
How about: `#error No known always_inline attribute for this platform.`

hebasto replied:
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/27575#issuecomment-1535337768).

theuni reviewed (changes requested):
Concept ACK.

hebasto commented:
Updated d66232f625a0781da2f4269cd8c5199ded2f43da -> a0d13e873be437260a79a62777560677c6a651b8 ([pr27575.01](https://github.com/hebasto/bitcoin/commits/pr27575.01) -> [pr27575.02](https://github.com/hebasto/bitcoin/commits/pr27575.02), [diff](https://github.com/hebasto/bitcoin/compare/pr27575.01..pr27575.02)):

- addressed @theuni's [comment](https://github.com/bitcoin/bitcoin/pull/27575#discussion_r1185412930)

hebasto committed:
Introduce platform-agnostic `ALWAYS_INLINE` macro

`<attributes.h>` has been included in anticipation of the following
commit.

hebasto committed:
scripted-diff: Use platform-agnostic `ALWAYS_INLINE` macro

-BEGIN VERIFY SCRIPT-
sed -i 's/ inline __attribute__((always_inline)) / ALWAYS_INLINE /g' $(git grep -l "inline __attribute__((always_inline))")
sed -i 's/ inline  __attribute__((always_inline)) / ALWAYS_INLINE /g' $(git grep -l "inline  __attribute__((always_inline))")
-END VERIFY SCRIPT-

hebasto commented:
Updated d66232f625a0781da2f4269cd8c5199ded2f43da -> 3f19875d667522412408d06873e87ff8150e49c4 ([pr27575.01](https://github.com/hebasto/bitcoin/commits/pr27575.01) -> [pr27575.03](https://github.com/hebasto/bitcoin/commits/pr27575.03), [diff](https://github.com/hebasto/bitcoin/compare/pr27575.01..pr27575.03)):

- addressed @theuni's [comment](https://github.com/bitcoin/bitcoin/pull/27575#discussion_r1185412930)

theuni reviewed (approved):
utACK 3f19875d667522412408d06873e87ff8150e49c4

fanquake reviewed (approved):
ACK 3f19875d667522412408d06873e87ff8150e49c4
